### PR TITLE
feat: add decompression syscall

### PR DESCRIPTION
**Pull request by ananas-block**
Changes:
- added solana altbn128 compression syscall.
  - it is added by copying it from solana_program v.1.17 for we currently cannot use solana_program v1.17 in light-protocol until we have upgraded to anchor 0.29.0
- added decompress_g1, decompress_g2 wrappers
- added tests:
  - bitmask: applying bit mask manually and decompressing g1, g2
  - e2e: decompressing and verifying a proof

### CODE REVIEW

File: src/syscalls/alt_bn128/compression.rs [hunk 1]
```diff
@@ -0,0 +1,601 @@
...
+    pub fn alt_bn128_g2_decompress(
+        g2_bytes: &[u8],
+    ) -> Result<[u8; alt_bn128_compression_size::G2], AltBn128CompressionError> {
+        let g2_bytes: [u8; alt_bn128_compression_size::G2_COMPRESSED] = g2_bytes
+            .try_into()
+            .map_err(|_| AltBn128CompressionError::InvalidInputSize)?;
+        if g2_bytes == [0u8; alt_bn128_compression_size::G2_COMPRESSED] {
+            return Ok([0u8; alt_bn128_compression_size::G2]);
+        }
+        println!("here");
```
Comment (vadorovsky) [inline src/syscalls/alt_bn128/compression.rs]: I guess it's some leftover? Remove all println-s, please :)

File: src/syscalls/alt_bn128/compression.rs [hunk 2]
```diff
@@ -0,0 +1,601 @@
...
+        let decompressed_g2 =
+            G2::deserialize_compressed(convert_endianness::<64, 64>(&g2_bytes).as_slice())
+                .map_err(|_| AltBn128CompressionError::G2DecompressionFailed)?;
+        println!("here");
```
Comment (vadorovsky) [inline src/syscalls/alt_bn128/compression.rs]: ditto

File: src/syscalls/alt_bn128/compression.rs [hunk 3]
```diff
@@ -0,0 +1,601 @@
...
+        let mut decompressed_g2_bytes = [0u8; alt_bn128_compression_size::G2];
+        decompressed_g2
+            .x
+            .serialize_with_mode(&mut decompressed_g2_bytes[..64], Compress::No)
+            .map_err(|_| AltBn128CompressionError::G2DecompressionFailed)?;
+        println!("here");
```
Comment (vadorovsky) [inline src/syscalls/alt_bn128/compression.rs]: ditto

File: src/syscalls/alt_bn128/compression.rs [hunk 4]
```diff
@@ -0,0 +1,601 @@
...
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::target_arch::convert_endianness;
+
+    use {
+        super::*,
+        // crate::convert_endianness,
```
Comment (vadorovsky) [inline src/syscalls/alt_bn128/compression.rs]: Unneeded comment

File: src/syscalls/alt_bn128/compression.rs [hunk 5]
```diff
@@ -0,0 +1,601 @@
...
+
+            assert_eq!(
+                alt_bn128_g1_compress(&decompressed).unwrap(),
+                compressed_ref
+            );
+            assert_eq!(decompressed, *g1_be);
+        }
+    }
+
+    #[ignore]
```
Comment (vadorovsky) [inline src/syscalls/alt_bn128/compression.rs]: Any reason why do we ignore this test? If there is some future work remaining, I would leave a TODO comment.

File: src/syscalls/alt_bn128/compression.rs [hunk 6]
```diff
@@ -0,0 +1,601 @@
...
+            let g1_neg_be: [u8; 64] = convert_endianness::<32, 64>(&g1_neg_be);
+
+            let points = [(g1, g1_be), (g1_neg, g1_neg_be)];
+
+            for (point, g1_be) in &points {
+                let mut compressed_ref = [0u8; 32];
+                G1::serialize_with_mode(point, compressed_ref.as_mut_slice(), Compress::Yes)
+                    .unwrap();
+                let compressed_ref: [u8; 32] = convert_endianness::<32, 32>(&compressed_ref);
+                // println!("point {:?}", point.to_flags());
```
Comment (vadorovsky) [inline src/syscalls/alt_bn128/compression.rs]: pls remove

**Review by vadorovsky — COMMENTED** [review]
Just couple of nits, otherwise looks good!

File: src/syscalls/alt_bn128/compression.rs (hunk 5 above)
Comment (ananas-block) [reply src/syscalls/alt_bn128/compression.rs]: It's too heavy to execute with this many iterations I will reduce the number of iterations.

Comment (ananas-block) [reply src/syscalls/alt_bn128/compression.rs]: reduced to 1k and 100 iterations so that execution time is bearable.

**Pull request merged**